util/grub-editenv.c

```diff
@@ -188,23 +190,116 @@ grub_zfs_fini (void *arg)
   libzfs_fini(g_zfs);
 }
 
-/* We need to convert ZFS's error returning pattern to the one we expect */
+typedef enum vbe_vers {
+        VB_RAW = 0,
+        VB_NVLIST = 1
+} vbe_vers_t;
+
+/*
+ * We need to convert ZFS's nvlist to a buffer we can manipulate in a
+ * platform-agnostic fashion.
+ */
+static int
+grub_zfs_get_bootenv (void *arg, char **buf, size_t *size)
+{
+  zpool_handle_t *zhp = arg;
+  nvlist_t *nvl;
+  int error = zpool_get_bootenv (zhp, &nvl);
+  nvpair_t *nvp = NULL;
+  size_t count;
+  u_int64_t version;
+  if (error == -1)
+    {
+      error = libzfs_errno (zpool_get_handle (zhp));
+      return error;
+    }
+  nvlist_size (nvl, size, NV_ENCODE_NATIVE);
+  *size *= 2;
+  if (*size < GRUB_ENVBLK_DEFAULT_SIZE)
+    *size = GRUB_ENVBLK_DEFAULT_SIZE;
+  *buf = malloc (*size);
+  if (nvlist_lookup_uint64 (nvl, "version", &version) != 0)
+    {
+      free (*buf);
+      nvlist_free (nvl);
+      return -1;
+    }
+  if (version == VB_RAW)
+    {
+      char *raw;
+      if (nvlist_lookup_string (nvl, "grub:envmap", &raw) != 0)
+        {
+          free (*buf);
+          nvlist_free (nvl);
+          return -1;
+        }
+      strncpy (*buf, raw, *size);
+      *size = strlen (raw);
+      nvlist_free (nvl);
+      return 0;
+    }
+  grub_util_create_envblk_buffer (*buf, *size);
+  count = (strchr (*buf, '\n') - *buf) + 1;
+  while ((nvp = nvlist_next_nvpair (nvl, nvp)) != NULL)
+    {
+      char *name, *value;
+      int rc;
+      name = nvpair_name (nvp);
+      if (nvpair_type (nvp) == DATA_TYPE_UINT64)
+        {
+          assert (strcmp(name, "version") == 0);
+          continue;
+        }
+      assert (nvpair_type (nvp) == DATA_TYPE_STRING);
+      nvpair_value_string (nvp, &value);
+      rc = snprintf (*buf + count, *size - count, "%s=%s\n", name, value);
+      if (rc < 0)
+        grub_util_error (_("cannot write nvlist entry %s to buffer"), name);
+      count += rc;
+      assert (count < *size);
```

Comment:
If the buffer is too small, count will be larger than size, and last entry would be truncated. Is this something we want to handle somehow with a meaningful error?

Reply:
Given that the buffer was allocated to be twice the size of the nvlist, I don't believe that's actually possible . I will add error handling for that case, though, just in case.